File: frontend/src/types/projectEvents.ts
export const PROJECT_EVENT_TYPES = [
  'pipeline_start',
  'chapter_writing',
  'chapter_truncated',
  'precheck_done',
  'prompt_generated',
  'prompt_generation_fallback',
  'prompt_updated',
  'content_chunk',
  'writer_non_stream_started',
  'writer_non_stream_succeeded',
  'writer_stream_no_text_fallback',
  'writer_stream_fallback_started',
  'writer_stream_fallback_succeeded',
  'chapter_reviewing',
  'chapter_scored',
  'memory_updated',
  'chapter_passed',
  'chapter_paused',
  'chapter_rewriting',
  'job_failed',
  'project_paused',
  'pipeline_complete',
] as const

export type ProjectEventType = (typeof PROJECT_EVENT_TYPES)[number]

export interface ProjectEventBase<TType extends string = string, TData extends object = Record<string, unknown>> {
  id: number
  type: TType
  chapter_number: number | null
  data: TData
  created_at: string
}

export interface PipelineStartEventData {
  chapter_number?: number
  resume?: boolean
}

export interface ChapterWritingEventData {
  attempt?: number
  continuation?: boolean
  continuation_source?: 'resume_saved_draft' | 'writer_truncated' | 'chapter_breakpoint_button'
  saved_chars?: number
}

export interface ChapterTruncatedEventData {
  attempt?: number
  accumulated_chars?: number
  reason?: string
}

export interface PromptGeneratedEventData {
  version_no?: number
  source?: string
  fallback?: Record<string, unknown>
}

export interface WriterTimeoutEventData {
  timeout_seconds?: number
}

export interface WriterFallbackEventData {
  waited_seconds?: number
  streamed_chars?: number
  fallback_chars?: number
  timeout_seconds?: number
  reason?: string
}

export interface ContentChunkEventData {
  chunk?: string
}

export interface ChapterScoredEventData {
  score?: number
  passed?: boolean
  review_thresholds?: Record<string, unknown>
}

export interface MemoryUpdatedEventData {
  chapter_number?: number
  character_revision_count?: number
  world_revision_count?: number
  applied_revision_count?: number
  needs_review_count?: number
  confidence_threshold?: number
}

export interface ChapterPassedEventData {
  score?: number | null
  manual?: boolean
  auto_accepted?: boolean
}

export interface ChapterPausedEventData {
  reason?: string
  stage?: string
  saved_chars?: number
}

export interface ChapterRewritingEventData {
  retry_count?: number
  queued?: boolean
  reason?: string
  error?: string
  failure_summary?: {
    overall_score?: number
    blocking_issue_count?: number
    top_blocking_issues?: string[]
    violated_instruction_count?: number
    top_violated_instructions?: string[]
    improvement_suggestion_count?: number
    top_improvement_suggestions?: string[]
  }
}

export interface JobFailedEventData {
  job_id?: string
  error?: string
}

export interface ProjectPausedEventData {
  reason?: string
  phase?: string
  cancelled_jobs?: number
  paused_chapters?: number
  next_chapter?: number
}

export interface PipelineCompleteEventData {
  project_id?: string
}

export interface ProjectEventDataMap {
  pipeline_start: PipelineStartEventData
  chapter_writing: ChapterWritingEventData
  chapter_truncated: ChapterTruncatedEventData
  precheck_done: Record<string, unknown>
  prompt_generated: PromptGeneratedEventData
  prompt_generation_fallback: Record<string, unknown>
  prompt_updated: Record<string, unknown>
  content_chunk: ContentChunkEventData
  writer_non_stream_started: WriterTimeoutEventData
  writer_non_stream_succeeded: WriterFallbackEventData
  writer_stream_no_text_fallback: WriterFallbackEventData
  writer_stream_fallback_started: WriterFallbackEventData
  writer_stream_fallback_succeeded: WriterFallbackEventData
  chapter_reviewing: Record<string, unknown>
  chapter_scored: ChapterScoredEventData
  memory_updated: MemoryUpdatedEventData
  chapter_passed: ChapterPassedEventData
  chapter_paused: ChapterPausedEventData
  chapter_rewriting: ChapterRewritingEventData
  job_failed: JobFailedEventData
  project_paused: ProjectPausedEventData
  pipeline_complete: PipelineCompleteEventData
}

export type KnownProjectEvent = {
  [Key in ProjectEventType]: ProjectEventBase<Key, ProjectEventDataMap[Key]>
}[ProjectEventType]

export type ProjectEvent = KnownProjectEvent | ProjectEventBase<string, Record<string, unknown>>

File: backend/services/project_events.py
import enum
import uuid
from typing import Any, TypedDict


class ProjectEventType(str, enum.Enum):
    PIPELINE_START = 'pipeline_start'
    CHAPTER_WRITING = 'chapter_writing'
    CHAPTER_TRUNCATED = 'chapter_truncated'
    PRECHECK_DONE = 'precheck_done'
    PROMPT_GENERATED = 'prompt_generated'
    PROMPT_GENERATION_FALLBACK = 'prompt_generation_fallback'
    PROMPT_UPDATED = 'prompt_updated'
    CONTENT_CHUNK = 'content_chunk'
    WRITER_NON_STREAM_STARTED = 'writer_non_stream_started'
    WRITER_NON_STREAM_SUCCEEDED = 'writer_non_stream_succeeded'
    WRITER_STREAM_NO_TEXT_FALLBACK = 'writer_stream_no_text_fallback'
    WRITER_STREAM_FALLBACK_STARTED = 'writer_stream_fallback_started'
    WRITER_STREAM_FALLBACK_SUCCEEDED = 'writer_stream_fallback_succeeded'
    CHAPTER_REVIEWING = 'chapter_reviewing'
    CHAPTER_SCORED = 'chapter_scored'
    MEMORY_UPDATED = 'memory_updated'
    CHAPTER_PASSED = 'chapter_passed'
    CHAPTER_PAUSED = 'chapter_paused'
    CHAPTER_REWRITING = 'chapter_rewriting'
    JOB_FAILED = 'job_failed'
    PROJECT_PAUSED = 'project_paused'
    PIPELINE_COMPLETE = 'pipeline_complete'


class PromptGeneratedPayload(TypedDict, total=False):
    version_no: int
    source: str
    fallback: dict[str, Any] | None


class MemoryUpdatedPayload(TypedDict, total=False):
    chapter_number: int
    character_revision_count: int
    world_revision_count: int
    applied_revision_count: int
    needs_review_count: int
    confidence_threshold: float


class ProjectPausedPayload(TypedDict, total=False):
    reason: str
    phase: str
    cancelled_jobs: int
    paused_chapters: int
    next_chapter: int


class ChapterRewriteFailureSummary(TypedDict, total=False):
    overall_score: float
    blocking_issue_count: int
    top_blocking_issues: list[str]
    violated_instruction_count: int
    top_violated_instructions: list[str]
    improvement_suggestion_count: int
    top_improvement_suggestions: list[str]


class ChapterTruncatedPayload(TypedDict, total=False):
    attempt: int
    accumulated_chars: int
    reason: str


class ChapterRewritingPayload(TypedDict, total=False):
    retry_count: int
    queued: bool
    reason: str
    error: str
    failure_summary: ChapterRewriteFailureSummary


class JobFailedPayload(TypedDict):
    job_id: str
    error: str


def pipeline_start_payload(*, chapter_number: int | None = None, resume: bool = False) -> dict[str, Any]:
    payload: dict[str, Any] = {}
    if chapter_number is not None:
        payload['chapter_number'] = chapter_number
    if resume:
        payload['resume'] = True
    return payload


def prompt_generated_payload(*, version_no: int, source: str, fallback: dict[str, Any] | None = None) -> PromptGeneratedPayload:
    payload: PromptGeneratedPayload = {
        'version_no': version_no,
        'source': source,
    }
    if fallback is not None:
        payload['fallback'] = fallback
    return payload


def memory_updated_payload(
    *,
    chapter_number: int,
    character_revision_count: int,
    world_revision_count: int,
    applied_revision_count: int,
    needs_review_count: int,
    confidence_threshold: float | None = None,
) -> MemoryUpdatedPayload:
    payload: MemoryUpdatedPayload = {
        'chapter_number': chapter_number,
        'character_revision_count': character_revision_count,
        'world_revision_count': world_revision_count,
        'applied_revision_count': applied_revision_count,
        'needs_review_count': needs_review_count,
    }
    if confidence_threshold is not None:
        payload['confidence_threshold'] = confidence_threshold
    return payload


def chapter_passed_payload(*, score: float | None, manual: bool = False, auto_accepted: bool = False) -> dict[str, Any]:
    payload: dict[str, Any] = {'score': score}
    if manual:
        payload['manual'] = True
    if auto_accepted:
        payload['auto_accepted'] = True
    return payload


def chapter_truncated_payload(
    *,
    attempt: int,
    accumulated_chars: int,
    reason: str,
) -> ChapterTruncatedPayload:
    return {
        'attempt': attempt,
        'accumulated_chars': accumulated_chars,
        'reason': reason,
    }


def chapter_rewriting_payload(
    *,
    retry_count: int | None = None,
    queued: bool = False,
    reason: str | None = None,
    error: str | None = None,
    failure_summary: ChapterRewriteFailureSummary | None = None,
) -> ChapterRewritingPayload:
    payload: ChapterRewritingPayload = {}
    if retry_count is not None:
        payload['retry_count'] = retry_count
    if queued:
        payload['queued'] = True
    if reason is not None:
        payload['reason'] = reason
    if error is not None:
        payload['error'] = error
    if failure_summary is not None:
        payload['failure_summary'] = failure_summary
    return payload


def project_paused_payload(
    *,
    reason: str,
    phase: str | None = None,
    cancelled_jobs: int | None = None,
    paused_chapters: int | None = None,
    next_chapter: int | None = None,
) -> ProjectPausedPayload:
    payload: ProjectPausedPayload = {'reason': reason}
    if phase is not None:
        payload['phase'] = phase
    if cancelled_jobs is not None:
        payload['cancelled_jobs'] = cancelled_jobs
    if paused_chapters is not None:
        payload['paused_chapters'] = paused_chapters
    if next_chapter is not None:
        payload['next_chapter'] = next_chapter
    return payload


def pipeline_complete_payload(project_id: uuid.UUID) -> dict[str, str]:
    return {'project_id': str(project_id)}


def job_failed_payload(*, job_id: uuid.UUID, error: str) -> JobFailedPayload:
    return {'job_id': str(job_id), 'error': error}
